From 584de2a3da974fdcdd6976f88339ab87558c3ede Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Thu, 8 Jun 2023 10:52:47 +0800
Subject: [PATCH 002/780] [Huawei] mm/dynamic_hugetlb: allow task migrate
 between cgroups which belong to the same dpool

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188703

--------------------------------

Allow task migrate between cgroups which belong to the same dpool.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 include/linux/dynamic_hugetlb.h |  5 +++--
 mm/dynamic_hugetlb.c            | 13 +++++++++++--
 mm/memcontrol.c                 |  2 +-
 3 files changed, 15 insertions(+), 5 deletions(-)

diff --git a/include/linux/dynamic_hugetlb.h b/include/linux/dynamic_hugetlb.h
index 9f42065890b2..5d6a2b77a5b1 100644
--- a/include/linux/dynamic_hugetlb.h
+++ b/include/linux/dynamic_hugetlb.h
@@ -123,7 +123,7 @@ struct page *alloc_page_from_dhugetlb_pool(gfp_t gfp, unsigned int order,
 					   unsigned int flags);
 bool free_page_to_dhugetlb_pool(struct page *page);
 void free_page_list_to_dhugetlb_pool(struct list_head *list);
-int task_has_mem_in_hpool(struct task_struct *tsk);
+int allow_task_migrate_in_hpool(struct task_struct *tsk, struct mem_cgroup *dst);
 
 void link_hpool(struct hugetlbfs_inode_info *p, struct hstate *h);
 void unlink_hpool(struct hugetlbfs_inode_info *p);
@@ -175,7 +175,8 @@ static inline bool free_page_to_dhugetlb_pool(struct page *page)
 static inline void free_page_list_to_dhugetlb_pool(struct list_head *list)
 {
 }
-static inline int task_has_mem_in_hpool(struct task_struct *tsk)
+static inline int allow_task_migrate_in_hpool(struct task_struct *tsk,
+					      struct mem_cgroup *dst)
 {
 	return 0;
 }
diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index 1babd868fe49..2e244427bd5b 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -535,7 +535,13 @@ static struct dhugetlb_pool *find_hpool_by_task(struct task_struct *tsk)
 	return memcg->hpool;
 }
 
-int task_has_mem_in_hpool(struct task_struct *tsk)
+/*
+ * Allow task migrate between cgroups if hpool is enabled iff one of the
+ * following conditions are met:
+ * a) current task does not belong to hpool
+ * b) from cgroup and dst cgroup belongs to the same hpool
+ */
+int allow_task_migrate_in_hpool(struct task_struct *tsk, struct mem_cgroup *dst)
 {
 	struct dhugetlb_pool *hpool;
 
@@ -544,7 +550,10 @@ int task_has_mem_in_hpool(struct task_struct *tsk)
 
 	hpool = find_hpool_by_task(tsk);
 
-	return hpool ? -EPERM : 0;
+	if (!hpool || (dst && hpool == dst->hpool))
+		return 0;
+
+	return -EPERM;
 }
 
 bool should_allocate_from_dhugetlb_pool(gfp_t gfp_mask)
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index de92ba7e65c8..3df0f6a40bd8 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -6537,7 +6537,7 @@ static int mem_cgroup_can_attach(struct cgroup_taskset *tset)
 	if (!p)
 		return 0;
 
-	ret = task_has_mem_in_hpool(p);
+	ret = allow_task_migrate_in_hpool(p, memcg);
 	if (ret)
 		return ret;
 
-- 
2.17.1

